# 지표 지도 그리기

## Funnel & North Star Framework

### 1. 분석 목적

고객이 프로모션을 수신한 뒤 열람하고 조건을 달성하기까지의 과정을 퍼널로 구성한다.

각 단계의 전환율을 확인하고, 프로모션 설계와 고객 특성에 따라 성과가 어떻게 달라지는지 분석한다.

---

## 2. North Star Metric

### 인센티브형 프로모션의 열람 후 완료율

> 열람 후 유효기간 내 완료된 BOGO·할인형 프로모션 수  
> ÷ 열람된 BOGO·할인형 프로모션 수 × 100

열람 후 완료율은 고객이 프로모션을 확인한 후 실제로 조건을 달성한 정도를 보여준다.

정보 제공형 프로모션은 보상과 달성 조건이 없어 완료 이벤트가 발생하지 않으므로 North Star Metric 계산에서 제외한다.

---

## 3. 프로모션 퍼널

| 퍼널 단계 | 지표 | 계산 기준 | 의미 |
|---|---|---|---|
| 수신 | 프로모션 수신 건수 | `offer received` 기록 수 | 고객에게 발송된 프로모션 규모 |
| 수신 → 열람 | 수신 후 열람률 | 열람된 프로모션 수 ÷ 수신된 프로모션 수 × 100 | 수신한 프로모션을 확인한 비율 |
| 열람 → 완료 | 열람 후 완료율 | 열람 후 완료된 인센티브형 프로모션 수 ÷ 열람된 인센티브형 프로모션 수 × 100 | 프로모션을 확인한 후 조건을 달성한 비율 |
| 수신 → 완료 | 프로모션 완료율 | 완료된 인센티브형 프로모션 수 ÷ 수신된 인센티브형 프로모션 수 × 100 | 수신한 프로모션 중 최종 완료된 비율 |
| 결제 | 결제금액 | `transaction`의 `amount` 합계 | 고객의 전체 구매 규모를 보여주는 보조 지표 |

인센티브형 프로모션은 BOGO와 할인형 프로모션을 의미한다.

---

## 4. 가설-지표 매핑

| 구분 | 가설 | 퍼널 단계 | 측정 지표 | 비교 기준 | 사용 데이터 | 측정 가능 여부 |
|---|---|---|---|---|---|---|
| 프로모션 유형 | 프로모션 유형에 따라 열람률에 차이가 있을 것이다 | 수신 → 열람 | 수신 후 열람률 | BOGO·할인·정보 제공형 | portfolio, transcript | 가능 |
| 프로모션 유형 | 할인형은 BOGO보다 완료율이 높을 것이다 | 수신 → 완료 | 프로모션 완료율 | 할인형·BOGO | portfolio, transcript | 가능 |
| 보상 | 보상 수준이 높을수록 완료율이 높을 것이다 | 열람 → 완료 | 열람 후 완료율 | 보상 수준 | portfolio, transcript | 가능 |
| 유효기간 | 유효기간이 길수록 완료율이 높을 것이다 | 열람 → 완료 | 열람 후 완료율 | 유효기간 | portfolio, transcript | 가능 |
| 난이도 | 달성 난이도가 높을수록 완료율이 낮을 것이다 | 열람 → 완료 | 열람 후 완료율 | 최소 결제 조건 | portfolio, transcript | 가능 |
| 채널 | 발송 채널 조합에 따라 열람률과 완료율에 차이가 있을 것이다 | 수신 → 열람, 열람 → 완료 | 수신 후 열람률, 열람 후 완료율 | 채널 포함 여부 및 조합 | portfolio, transcript | 제한적으로 가능 |
| 성별 | 여성은 남성보다 완료율이 높을 것이다 | 열람 → 완료 | 열람 후 완료율 | 여성·남성 | profile, transcript | 가능 |
| 소득 | 소득이 높은 고객은 완료율이 높을 것이다 | 열람 → 완료 | 열람 후 완료율 | 소득 구간 | profile, transcript | 가능 |
| 연령 | 연령이 높은 고객은 완료율이 낮을 것이다 | 열람 → 완료 | 열람 후 완료율 | 연령대 | profile, transcript | 가능 |

---

## 5. 지표 계산 기준

### 수신 후 열람률

> 열람된 프로모션 수 ÷ 수신된 프로모션 수 × 100

BOGO·할인·정보 제공형을 모두 포함해 계산할 수 있다.

### 프로모션 완료율

> 완료된 인센티브형 프로모션 수 ÷ 수신된 인센티브형 프로모션 수 × 100

완료 조건이 있는 BOGO와 할인형만 포함한다.

### 열람 후 완료율

> 열람 후 완료된 인센티브형 프로모션 수 ÷ 열람된 인센티브형 프로모션 수 × 100

열람 시점보다 완료 시점이 늦고, 프로모션 유효기간 안에 완료된 경우만 포함한다.

---

## 6. 비교 구간 초안

가설 검정에 사용할 구간은 팀 논의를 통해 최종 확정한다.

현재 데이터에서 확인한 기준값은 다음과 같다.

| 항목 | 확인된 기준값 |
|---|---:|
| 보상 평균 | 4.2 |
| 유효기간 평균 | 6.5일 |
| 난이도 평균 | 7.7 |
| 고객 평균 나이 | 약 54.39세 |
| 고객 소득 중앙값 | 64,000달러 |
| 고객 소득 평균 | 약 65,405달러 |

- 연령을 평균 기준으로 구분하면 54세 이하와 55세oria 이상으로 나눌 수 있다.
- 소득 8만 달러 이상을 고소득 고객으로 설정할 경우, 데이터BELMelissa 평균이 아니라 팀이 정한 구간임을 명시한다.

---

## 7. 측정 시 주의사항

- 동일 고객이 같은 프로모션을 여러 번 받을 수 있으므로 고객·프로모션·수신 시점을 기준으로 각 프로모션 수신 건을 구분해야 한다.
- 열람과 완료는 해당 프로모션의 수신 이후이면서 유효기간 안에 발생한 경우만 연결해야 한다.
- 정보 제공형 프로모션은 완료 조건이 없으므로 열람률까지만 평가한다.
- 여러 채널이 하나의 프로모션에 함께 포함되어 있어 개별 채널의 독립적인 효과를 확정하기 어렵다.
- 채널 분석은 채널의 인과효과가 아니라 채널 포함 여부 또는 채널 조합과 성과의 연관성으로 해석한다.
- 고객정보가 누락되고 나이가 118세로 기록된 2,175명은 고객 특성별 분석에서 별도로 처리한다.
- 결제 기록에는 프로모션 ID가 없으므로 특정 결제를 특정 프로모션의 성과로 직접 귀속할 수 없다.
- 프로모션 설계는 10개뿐이므로 보상·기간·난이도·채널 각각의 독립적인 효과를 강하게 단정하지 않는다.

---

## 8. 현재 측정하기 어려운 지표

| 지표 | 측정이 어려운 이유 | 추가로 필요한 데이터 |
|---|---|---|
| 개별 채널의 직접 효과 | 실제로 어떤 채널에서 열람했는지 알 수 없음 | 채널별 노출·클릭 로그 |
| 프로모션의 정확한 구매 유발 효과 | 결제 기록에 프로모션 ID가 없음 | 결제별 프로모션 연결 정보 또는 대조군 |
| 정확한 ROI | 프로모션 운영비와 상품 원가 정보가 없음 | 캠페인 비용·상품 원가·마진 데이터 |
| 장기 재구매 효과 | 관찰기간이 약 30일로 짧음 | 장기간 고객 행동 기록 |

## 지표 계산


In [1]:
import pandas as pd
from pathlib import Path

DATA_DIR = Path("raw")

portfolio = pd.read_json(
    DATA_DIR / "portfolio.json",
    lines=True
)

profile = pd.read_json(
    DATA_DIR / "profile.json",
    lines=True
)

transcript = pd.read_json(
    DATA_DIR / "transcript.json",
    lines=True
)

print(portfolio.shape)
print(profile.shape)
print(transcript.shape)

(10, 6)
(17000, 5)
(306534, 4)


In [2]:
transcript_work = transcript.copy()

transcript_work["offer_id"] = transcript_work["value"].apply(
    lambda x: x.get("offer id", x.get("offer_id"))
)

display(
    transcript_work[
        ["person", "event", "offer_id", "time"]
    ].head()
)

,person,event,offer_id,time
0,78afa995795e4d85b5d9ceeca43f5fef,offer received,9b98b8c7a33c4b65b9aebfe6a799e6d9,0
1,a03223e636434f42ac4c3df47e8bac43,offer received,0b1e1539f2cc45b7b9fa7c272da2e1d7,0
2,e2127556f4f64592b11af22de27a7932,offer received,2906b810c7d4411798c6938adc9daaa5,0
3,8ec6ce2a7e7949b1bf142def7d0e0586,offer received,fafdcd668e3743c1bb461111dcafc2a4,0
4,68617ca6246f4fbc85e91a2a49552598,offer received,4d5c57ea9a6940dd891ad53e9dbe8da0,0


In [6]:
offer_id_check = (
    transcript_work
    .groupby("event")["offer_id"]
    .agg(
        전체_기록="size",
        ID_존재="count"
    )
)

display(offer_id_check)

,전체_기록,ID_존재
event,,
offer completed,33579,33579
offer received,76277,76277
offer viewed,57725,57725
transaction,138953,0


### 추출 결과

- 프로모션 수신 76,277건, 열람 57,725건, 완료 33,579건에서 프로모션 ID가 모두 정상적으로 추출되었다.
- 결제 138,953건에는 원본 데이터상 프로모션 ID가 없으므로 `offer_id`가 결측치로 남는 것이 정상이다.
- 프로모션 관련 기록에서 ID가 누락된 경우는 확인되지 않았다.

In [3]:
# 프로모션 유효기간을 시간 단위로 변환
duration_hours = (
    portfolio.set_index("id")["duration"] * 24
).to_dict()

# 결제를 제외한 프로모션 행동만 선택
offer_events = transcript_work.loc[
    transcript_work["event"] != "transaction",
    ["person", "event", "time", "offer_id"]
].copy()

# 같은 시간에는 수신 → 열람 → 완료 순서로 처리
event_order = {
    "offer received": 0,
    "offer viewed": 1,
    "offer completed": 2
}

offer_events["event_order"] = (
    offer_events["event"].map(event_order)
)

offer_events = offer_events.sort_values(
    ["person", "offer_id", "time", "event_order"]
)

offer_instances = []

for (person, offer_id), group in offer_events.groupby(
    ["person", "offer_id"],
    sort=False
):
    received_instances = []

    for row in group.itertuples(index=False):

        if row.event == "offer received":
            received_instances.append({
                "customer_id": person,
                "offer_id": offer_id,
                "received_time": row.time,
                "expiry_time": (
                    row.time + duration_hours[offer_id]
                ),
                "viewed_time": None,
                "completed_time": None
            })

        elif row.event == "offer viewed":
            candidates = [
                instance
                for instance in received_instances
                if (
                    instance["received_time"]
                    <= row.time
                    <= instance["expiry_time"]
                    and instance["viewed_time"] is None
                )
            ]

            if candidates:
                latest_instance = max(
                    candidates,
                    key=lambda x: x["received_time"]
                )
                latest_instance["viewed_time"] = row.time

        elif row.event == "offer completed":
            candidates = [
                instance
                for instance in received_instances
                if (
                    instance["received_time"]
                    <= row.time
                    <= instance["expiry_time"]
                    and instance["completed_time"] is None
                )
            ]

            if candidates:
                latest_instance = max(
                    candidates,
                    key=lambda x: x["received_time"]
                )
                latest_instance["completed_time"] = row.time

    offer_instances.extend(received_instances)

offer_funnel = pd.DataFrame(offer_instances)

print("생성된 프로모션 수신 건:", len(offer_funnel))

생성된 프로모션 수신 건: 76277


### 프로모션 수신 단위 생성 결과

고객별 프로모션 성과를 계산하기 위해 행동 기록을 프로모션 수신 건 단위로 재구성했다.

- 원본 프로모션 수신 기록: 76,277건
- 생성된 프로모션 수신 단위: 76,277건
- 누락되거나 추가된 수신 기록: 0건

각 수신 건에는 고객 ID, 프로모션 ID, 수신 시점과 유효기간을 부여했다. 이후 유효기간 안에 발생한 열람 및 완료 기록을 해당 수신 건에 연결한다.

동일 고객이 동일한 프로모션을 여러 번 받은 경우에는 행동 발생 시점을 기준으로 유효한 가장 최근 수신 건에 연결하는 잠정 기준을 사용했다. 최종 분석 전 연결되지 않은 열람·완료 기록과 중복 기록을 추가로 점검해야 한다.

In [4]:
offer_funnel = offer_funnel.merge(
    portfolio[
        [
            "id",
            "offer_type",
            "reward",
            "difficulty",
            "duration",
            "channels"
        ]
    ],
    left_on="offer_id",
    right_on="id",
    how="left"
).drop(columns="id")

offer_funnel["viewed"] = (
    offer_funnel["viewed_time"].notna()
)

offer_funnel["completed"] = (
    offer_funnel["completed_time"].notna()
)

offer_funnel["viewed_before_completed"] = (
    offer_funnel["viewed"]
    & offer_funnel["completed"]
    & (
        offer_funnel["viewed_time"]
        <= offer_funnel["completed_time"]
    )
)

offer_funnel["completed_without_prior_view"] = (
    offer_funnel["completed"]
    & ~offer_funnel["viewed_before_completed"]
)

display(offer_funnel.head())

,customer_id,offer_id,received_time,expiry_time,viewed_time,completed_time,offer_type,reward,difficulty,duration,channels,viewed,completed,viewed_before_completed,completed_without_prior_view
0,0009655768c64bdeb2e877511632db8f,2906b810c7d4411798c6938adc9daaa5,576,744,NaN,576.0,discount,2,10,7,"[web, email, mobile]",False,True,False,True
1,0009655768c64bdeb2e877511632db8f,3f207df678b143eea3cee63160fa8bed,336,432,372.0,NaN,informational,0,0,4,"[web, email, mobile]",True,False,False,False
2,0009655768c64bdeb2e877511632db8f,5a8bc65990b245e5a138643cd4eb9837,168,240,192.0,NaN,informational,0,0,3,"[email, mobile, social]",True,False,False,False
3,0009655768c64bdeb2e877511632db8f,f19421c1d4aa40978ebb69ca19b0e20d,408,528,456.0,414.0,bogo,5,5,5,"[web, email, mobile, social]",True,True,False,True
4,0009655768c64bdeb2e877511632db8f,fafdcd668e3743c1bb461111dcafc2a4,504,744,540.0,528.0,discount,2,10,10,"[web, email, mobile, social]",True,True,False,True


In [5]:
incentive_mask = offer_funnel["offer_type"].isin(
    ["bogo", "discount"]
)

all_received = len(offer_funnel)
all_viewed = offer_funnel["viewed"].sum()

incentive_received = incentive_mask.sum()
incentive_viewed = (
    incentive_mask & offer_funnel["viewed"]
).sum()

incentive_completed = (
    incentive_mask & offer_funnel["completed"]
).sum()

viewed_then_completed = (
    incentive_mask
    & offer_funnel["viewed_before_completed"]
).sum()

completed_without_view = (
    incentive_mask
    & offer_funnel["completed_without_prior_view"]
).sum()

print("[전체 프로모션]")
print("수신 건수:", all_received)
print("열람 건수:", all_viewed)
print("수신 후 열람률:", round(all_viewed / all_received * 100, 2), "%")

print("\n[BOGO·할인형 프로모션]")
print("수신 건수:", incentive_received)
print("열람 건수:", incentive_viewed)
print("완료 건수:", incentive_completed)

print(
    "수신 후 열람률:",
    round(incentive_viewed / incentive_received * 100, 2),
    "%"
)

print(
    "수신 후 완료율:",
    round(incentive_completed / incentive_received * 100, 2),
    "%"
)

print(
    "열람 후 완료율:",
    round(viewed_then_completed / incentive_viewed * 100, 2),
    "%"
)

print(
    "완료 중 사전 열람이 없었던 비율:",
    round(completed_without_view / incentive_completed * 100, 2),
    "%"
)

[전체 프로모션]
수신 건수: 76277
열람 건수: 56567
수신 후 열람률: 74.16 %

[BOGO·할인형 프로모션]
수신 건수: 61042
열람 건수: 46620
완료 건수: 33549
수신 후 열람률: 76.37 %
수신 후 완료율: 54.96 %
열람 후 완료율: 50.53 %
완료 중 사전 열람이 없었던 비율: 29.79 %
